Public/Src/Cache/ContentStore/Distributed/NuCache/FlushableCache.cs

```diff
@@ -0,0 +1,147 @@
+    /// that class.
+    /// </summary>
+    internal class FlushableCache
+    {
+        private readonly FlushableCacheConfiguration _configuration;
+        private readonly ContentLocationDatabase _database;
+
+        private ConcurrentBigMap<ShortHash, ContentLocationEntry> _cache = new ConcurrentBigMap<ShortHash, ContentLocationEntry>();
+        private ConcurrentBigMap<ShortHash, ContentLocationEntry> _flushingCache = new ConcurrentBigMap<ShortHash, ContentLocationEntry>();
+
+        private readonly SemaphoreSlim _flushTurnstile = new SemaphoreSlim(1);
+        private readonly ReadWriteLock _exchangeLock = ReadWriteLock.Create();
+
+        public FlushableCache(FlushableCacheConfiguration configuration, ContentLocationDatabase database)
+        {
+            _configuration = configuration;
+            _database = database;
+        }
+
+        /// <nodoc />
+        public void UnsafeClear()
+        {
+            // Order is important here, inverse order could cause deadlock.
+            using (_flushTurnstile.AcquireSemaphore())
+            using (_exchangeLock.AcquireWriteLock())
+            {
+                if (_cache.Count != 0)
+                {
+                    // Nothing guarantees that some number of updates couldn't have happened in between the last flush
+                    // and this reset, because acquisition of the write lock happens after the flush lock. The only way
+                    // to deal with this situation is to force a flush before this assignment and after locks have been
+                    // acquired. However, this isn't required by our code right now; although it is a simple change.
+                    _cache = new ConcurrentBigMap<ShortHash, ContentLocationEntry>();
+                }
+
+                // There should be no flushes ongoing, or we wouldn't have acquired the lock.
+                Contract.Assert(_flushingCache.Count == 0);
+            }
+        }
+
+        /// <nodoc />
+        public void Store(OperationContext context, ShortHash hash, ContentLocationEntry entry)
+        {
+            using (_exchangeLock.AcquireReadLock())
+            {
+                _cache[hash] = entry;
+            }
+        }
+
+        /// <nodoc />
+        public bool TryGetEntry(ShortHash hash, out ContentLocationEntry entry)
+        {
+            using (_exchangeLock.AcquireReadLock())
+            {
+                // The entry could be a tombstone, so we need to make sure the user knows content has actually been
+                // deleted, which is why we check for null.
+                if (_cache.TryGetValue(hash, out entry))
+                {
+                    _database.Counters[ContentLocationDatabaseCounters.TotalNumberOfCacheHit].Increment();
+                    return entry != null;
+                }
+                else if (_flushingCache.TryGetValue(hash, out entry))

```

Comment:
>_flushingCache.TryGetValue(hash, out entry) [](start = 25, length = 43)

Only need to check flushingCache while flushing. Add boolean _isFlushing and set/unset it before and after flushing.

Reply:
No longer applies due to latest changes, as discussed in person

---
In reply to: [291623570](https://github.com/microsoft/BuildXL/pull/403#discussion_r291623570) [](ancestors = 291623570)